src/Shared/Components/PooledByteBufferWritter.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Components.Infrastructure
+{
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        private byte[] _currentBuffer;
+        private int _index;
+        private bool _owned;
+
+        private const int MinimumBufferSize = 256;
+
+        public PooledByteBufferWriter(int initialCapacity = 0)
+        {
+            _currentBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);
+            _index = 0;
+        }
+
+        public PooledByteBufferWriter(byte[] existingBuffer)
+        {
+            _currentBuffer = existingBuffer;
+            _index = existingBuffer.Length;
+            _owned = true;
+        }
+
+        public ReadOnlyMemory<byte> WrittenMemory
+        {
+            get
+            {
+                return _currentBuffer.AsMemory(0, _index);
+            }
+        }
+
+        public int WrittenCount
+        {
+            get
+            {
+                return _index;
+            }
+        }
+
+        public int Capacity
+        {
+            get
+            {
+                return _currentBuffer.Length;
+            }
+        }
+
+        public int FreeCapacity
+        {
+            get
+            {
+                return _currentBuffer.Length - _index;
+            }
+        }
+
+        public void Clear()
+        {
+            ClearHelper();
+        }
+
+        private void ClearHelper()
+        {
+            _currentBuffer.AsSpan(0, _index).Clear();
+            _index = 0;
+        }
+
+        // Returns the rented buffer back to the pool
+        public void Dispose()
+        {
+            if (_currentBuffer == null || _owned)
+            {
+                return;
+            }
+
+            ClearHelper();
+            ArrayPool<byte>.Shared.Return(_currentBuffer);
+            _currentBuffer = null!;
```

Comment:
For thread safety, it might be better to make sure we've nulled out the `_currentBuffer` field before returning to the pool, otherwise it's possible that some other thread calls a "write" method on this instance which continues trying to write to `_currentBuffer` which will then be used in some other user's circuit.

Reply:
done